spec/models/journal_notification_mailer_spec.rb

```diff
@@ -0,0 +1,138 @@
+  end
+  let(:work_package) do
+    FactoryGirl.create(:work_package,
+                       project_id: project.id,
+                       project: project,
+                       author: user,
+                       type: project.types.first)
+  end
+  let(:notifications) { [] }
+
+  before do
+    allow(User).to receive(:current).and_return(user)
+    allow(Setting).to receive(:notified_events).and_return(notifications)
+
+    ActionMailer::Base.deliveries.clear
+  end
+
+  shared_examples_for 'handles deliveries' do |notification_setting|
+    context 'sends a notification' do
+      let(:notifications) { [notification_setting] }
+
+      it do
+        expect(ActionMailer::Base.deliveries.size).to eq(1)
+      end
+    end
+
+    it 'sends no notification' do
+      expect(ActionMailer::Base.deliveries.size).to eq(0)
+    end
+  end
+
+  describe 'journal creation' do
+    context 'work_package_created' do
+      before do
+        FactoryGirl.create(:work_package, project: project)
+      end
+
+      context 'sends a notification' do
+        let(:notifications) { ['work_package_added'] }
+
+        it do
+          expect(ActionMailer::Base.deliveries.size).to eq(1)
+        end
+      end
+
+      it 'sends no notification' do
+        expect(ActionMailer::Base.deliveries.size).to eq(0)
+      end
+    end
+
+    context 'work_package_updated' do
+      before do
+        work_package.add_journal(user)
+        work_package.subject = 'A change to the issue'
+        expect(work_package.save(validate: false)).to be_truthy
+      end
+
+      it_behaves_like 'handles deliveries', 'work_package_updated'
+    end
+
+    context 'work_package_note_added' do
+      before do
+        work_package.add_journal(user, 'This update has a note')
+        expect(work_package.save(validate: false)).to be_truthy
+        work_package.recreate_initial_journal!
+      end
+
+      it_behaves_like 'handles deliveries', 'work_package_note_added'
+    end
+
+    context 'status_updated' do
+      before do
+        work_package.add_journal(user)
+        work_package.status = FactoryGirl.build(:status)
+        expect(work_package.save(validate: false)).to be_truthy
+      end
+
+      it_behaves_like 'handles deliveries', 'status_updated'
+    end
+
+    context 'work_package_priority_updated' do
+      before do
+        work_package.add_journal(user)
+        work_package.priority = IssuePriority.generate!
+        expect(work_package.save(validate: false)).to be_truthy
+      end
+
+      it_behaves_like 'handles deliveries', 'work_package_priority_updated'
+    end
+
+    context 'send_notification disabled' do
+      before do
+        allow(JournalManager).to receive(:send_notification).and_return(false)
+        FactoryGirl.create(:work_package, project: project) # Provoke notification
+      end
+
+      it 'sends no notification' do
+        expect(ActionMailer::Base.deliveries.size).to eq(0)
+      end
+    end
+  end
+end
```

Comment:
The specs here look fine at a basic level and are probably sufficient. However, what makes me sad is that I can make all of them pass by changing the implementation in the following wrong way:

````ruby
def send_update_notification?(journal)
  Setting.notified_events.include?('work_package_updated') ||
    Setting.notified_events.include?('work_package_note_added') ||
    Setting.notified_events.include?('status_updated') ||
    Setting.notified_events.include?('work_package_priority_updated')
end
````

The problem is that all your specs work in the same way:

1. mock a WP so that it will fire for the condition under test
2. enable the condition under test (expect it fires)
3. disable all conditions (expect nothing fires)

What you really should do is:

1. enable the condition under test
2. mock a WP so that it will fire for that condition (expect it fires)
3. mock a WP so that it will not fire for that condition (expect nothing fires)
4. disable all conditions, use a mocked WP like in step 2 (expect that still nothing fires)